lib/internal/modules/cjs/loader.js

```diff
@@ -597,14 +600,28 @@ Module._resolveLookupPaths = function(request, parent, newReturn) {
 //    Then have it load  the file contents before returning its exports
 //    object.
 Module._load = function(request, parent, isMain) {
+  let relResolveCacheIdentifier;
   if (parent) {
     debug('Module._load REQUEST %s parent: %s', request, parent.id);
+    // Fast path for (lazy loaded) modules in the same directory. The indirect
+    // caching is required to allow cache invalidation without changing the old
+    // cache key names.
+    relResolveCacheIdentifier = `${parent.path}\x00${request}`;
+    const filename = relativeResolveCache[relResolveCacheIdentifier];
+    if (filename !== undefined) {
+      const cachedModule = Module._cache[filename];
+      if (cachedModule !== undefined) {
+        updateChildren(parent, cachedModule, true);
+        return cachedModule.exports;
+      }
+      delete relativeResolveCache[relResolveCacheIdentifier];
+    }
   }
 
-  var filename = Module._resolveFilename(request, parent, isMain);
+  const filename = Module._resolveFilename(request, parent, isMain);
 
-  var cachedModule = Module._cache[filename];
-  if (cachedModule) {
+  const cachedModule = Module._cache[filename];
+  if (cachedModule !== undefined) {
     updateChildren(parent, cachedModule, true);
     return cachedModule.exports;
   }
```

Comment:
Could this be done at the `Module._resolveFilename` level?
A cache key could be something like
```js
cacheKey =
  request + "\0" +
  dirPath + "\0" +
  (isMain ? "1" : "")
```

Reply:
That would not be beneficial as far as I can tell. I would also rather not move any parts around since the loader is highly monkey patched and every subtle change can cause side-effects.

`isMain` should not have any impact on the key as it's only about the very first require call and that is not required for the cache key.